// parallel-park/dist/child-process-worker.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const vm_1 = __importDefault(require("vm"));
const make_module_env_1 = __importDefault(require("make-module-env"));
const comms = fs_1.default.createWriteStream(
// @ts-ignore
null, { fd: 3 });
const [inputsString, fnString, callingFile] = process.argv.slice(2);
function onSuccess(data) {
    comms.write(JSON.stringify({ type: "success", data }));
}
function onError(error) {
    comms.write(JSON.stringify({
        type: "error",
        error: {
            name: error.name,
            message: error.message,
            stack: error.stack,
        },
    }));
}
try {
    const wrapperFn = vm_1.default.runInThisContext(`(function moduleWrapper(exports, require, module, __filename, __dirname) {
return ${fnString};})`);
    const inputs = JSON.parse(inputsString);
    const env = (0, make_module_env_1.default)(callingFile);
    const fn = wrapperFn(env.exports, env.require, env.module, env.__filename, env.__dirname);
    const result = fn(inputs);
    if (typeof result === "object" &&
        result != null &&
        typeof result.then === "function") {
        result.then(onSuccess, onError);
    }
    else {
        onSuccess(result);
    }
}
catch (err) {
    onError(err);
}
